src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs

```diff
@@ -1142,6 +1188,84 @@ protected virtual void ReportUnassigned(Symbol symbol, SyntaxNode node, int slot
 
             // mark the variable's slot so that we don't complain about the variable again
             _alreadyReported[slot] = true;
+
+            bool doesNotNeedImplicitFieldInitialization()
+                => !_requireOutParamsAssigned
+                || CurrentSymbol is not MethodSymbol { MethodKind: MethodKind.Constructor, ContainingType.TypeKind: TypeKind.Struct };
+
+            ErrorCode adjustErrorForStructThis(Symbol thisParameter, int thisSlot)
+            {
+                if (doesNotNeedImplicitFieldInitialization())
+                {
+                    return ErrorCode.ERR_UseDefViolationThis;
+                }
+
+                foreach (var field in _emptyStructTypeCache.GetStructInstanceFields(thisParameter.ContainingType))
+                {
+                    if (_emptyStructTypeCache.IsEmptyStructType(field.Type))
+                        continue;
+
+                    if (field is TupleErrorFieldSymbol)
+                        continue;
+
+                    int slot = VariableSlot(field, thisSlot);
+                    if (slot == -1 || !State.IsAssigned(slot))
+                    {
+                        Debug.Assert((object)field.ContainingType == CurrentSymbol.EnclosingThisSymbol().Type);
+                        AddImplicitlyInitializedField(field);
+                    }
+                }
+
+                if (!compilation.IsFeatureEnabled(MessageID.IDS_FeatureImplicitInitializationInStructConstructors))
+                {
+                    return ErrorCode.ERR_UseDefViolationThis;
+                }
+
+                return ErrorCode.WRN_UseDefViolationStructThis;
+            }
+
+            ErrorCode adjustErrorForStructFieldOrProperty(bool isProperty, int fieldSlot)
+            {
+                if (doesNotNeedImplicitFieldInitialization())
+                {
+                    return isProperty ? ErrorCode.ERR_UseDefViolationProperty : ErrorCode.ERR_UseDefViolationField;
+                }
+
+                var thisSlot = GetOrCreateSlot(CurrentSymbol.EnclosingThisSymbol(), createIfMissing: false);
+                var localSlot = fieldSlot;
+                while (localSlot != -1 && RootSlot(localSlot) is int root && root != thisSlot)
```

Comment:
Let's chat on Monday. I need some help walking through this.

Reply:
We spoke offline. This code was wrong. I was able to simplify the implementation once I changed away from calling 'RootSlot' here.

Reply:
Looks much better now. Thanks